File: kora-app-annotation-processor/src/test/java/io/koraframework/kora/app/annotation/processor/app/AppWithFactories4.java
package io.koraframework.kora.app.annotation.processor.app;

import io.koraframework.application.graph.TypeRef;
import io.koraframework.common.KoraApp;
import io.koraframework.common.annotation.Root;

import java.util.ArrayList;
import java.util.List;

@KoraApp
public interface AppWithFactories4 {

    default Integer intComponent() {
        return 0;
    }

    default <T extends List<Class1>> TwoGenericClass<T, Class2> factory(TypeRef<T> typeRef1, TypeRef<Class2> typeRef2, Integer dependency) {
        return new TwoGenericClass<>();
    }

    @Root
    default Class1 class1(TwoGenericClass<ArrayList<Class1>, Class2> genericClass) {
        return new Class1();
    }

    class TwoGenericClass<T, Q> {}

    class Class1 {}

    class Class2 {}
}

File: kora-app-symbol-processor/src/test/kotlin/io/koraframework/kora/app/ksp/app/AppWithFactories4.kt
package io.koraframework.kora.app.ksp.app

import io.koraframework.application.graph.TypeRef
import io.koraframework.common.KoraApp
import io.koraframework.common.annotation.Root

@KoraApp
interface AppWithFactories4 {
    fun intComponent(): Int {
        return 0
    }

    fun <T : List<Class1>> factory(
        typeRef1: TypeRef<T>,
        typeRef2: TypeRef<Class2>,
        dependency: Int
    ): TwoGenericClass<T, Class2> {
        return TwoGenericClass()
    }

    @Root
    fun class1(genericClass: TwoGenericClass<List<Class1>, Class2>): Class1 {
        return Class1()
    }

    class TwoGenericClass<T, Q>
    class Class1
    class Class2
}
